drivers/sensor/bq34110/bq34110.c

```diff
@@ -0,0 +1,380 @@
+					  regs[i].dest);
+		if (rc != 0) {
+			LOG_ERR("Failed to read channel %d", chan);
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+static int bq34110_ctrl_reg_read(const struct device *dev,
+				 uint16_t subcommand,
+				 uint16_t *val, uint8_t read_addr)
+{
+	int ret;
+	uint8_t i2c_write_data[3], read_data[2], read_reg;
+	const struct bq34110_config *bq34110_cfg = dev->config;
+
+	memset(&i2c_write_data, 0, sizeof(i2c_write_data));
+	i2c_write_data[0] = MANUFACTURER_ACCESS_CONTROL;
+	i2c_write_data[1] = (uint8_t)(subcommand & 0x00FF);
+	i2c_write_data[2] = (uint8_t)((subcommand >> 8) & 0x00FF);
+	ret = i2c_write(bq34110_cfg->i2c_dev, i2c_write_data,
+			sizeof(i2c_write_data), bq34110_cfg->i2c_addr);
+	if (ret < 0) {
+		LOG_ERR("Failed to write");
+		return ret;
+	}
+
+	k_msleep(BQ34110_DELAY);
+
+	read_reg = read_addr;
+	ret = i2c_write_read(bq34110_cfg->i2c_dev, bq34110_cfg->i2c_addr,
+			     &read_reg, sizeof(read_reg), &read_data,
+			     sizeof(read_data));
+	if (ret < 0) {
+		LOG_ERR("Failed to read write");
+		return ret;
+	}
+
+	*val = ((read_data[1] << 8) | read_data[0]);
+
+	return 0;
+}
+
+static int updateDFparameter(const struct device *dev,
+			uint16_t df_addr, uint8_t data[], uint8_t datalength)
+{
+	int ret;
+	uint16_t datasum;
+	uint8_t i, i2c_df_addr_data[3],
+		i2c_datasum_datalen[3],
+		i2c_df_write_data[datalength+1];
+	const struct bq34110_config *bq34110_cfg = dev->config;
+
+	i2c_df_addr_data[0] = MANUFACTURER_ACCESS_CONTROL;
+	i2c_df_addr_data[1] = (uint8_t)(df_addr & 0x00FF);
+	i2c_df_addr_data[2] = (uint8_t)((df_addr >> 8) & 0x00FF);
+
+	ret = i2c_write(bq34110_cfg->i2c_dev, i2c_df_addr_data,
+			sizeof(i2c_df_addr_data), bq34110_cfg->i2c_addr);
+	if (ret < 0) {
+		LOG_ERR("Failed to write MAC");
+		return ret;
+	}
+
+	i2c_df_write_data[0] = MAC_DATA;
+	for (i = 1; i <= datalength; i++) {
+		i2c_df_write_data[i] = data[i-1];
+	}
+
+	ret = i2c_write(bq34110_cfg->i2c_dev, i2c_df_write_data,
+			sizeof(i2c_df_write_data), bq34110_cfg->i2c_addr);
+	if (ret < 0) {
+		LOG_ERR("Failed to write MAC");
+		return ret;
+	}
+
+	datasum = i2c_df_addr_data[1] + i2c_df_addr_data[2];
+	for (i = 1; i <= datalength; i++) {
+		datasum += i2c_df_write_data[i];
+	}
+
+	i2c_datasum_datalen[0] = MAC_DATA_SUM;
+	i2c_datasum_datalen[1] = (uint8_t)~datasum;
+	i2c_datasum_datalen[2] = 4 + datalength;
+	ret = i2c_write(bq34110_cfg->i2c_dev, i2c_datasum_datalen,
+			sizeof(i2c_datasum_datalen), bq34110_cfg->i2c_addr);
+	if (ret < 0) {
+		LOG_ERR("Failed to write MACDatasum");
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief initialise the fuel gauge
+ *
+ * @return 0 for success
+ */
+static int bq34110_gauge_init(const struct device *dev)
+{
+	int ret;
+	uint8_t writedata[2];
+	int16_t design_capacity;
+	const struct bq34110_config *bq34110_cfg = dev->config;
+	uint16_t design_voltage, device_type, taper_current, taper_voltage;
+
+	if (!device_is_ready(bq34110_cfg->i2c_dev)) {
+		LOG_ERR("bq34110 device is not ready");
```

Comment:
This is a bit misleading because it's the bus that's not ready, not the fuel gauge. Please adjust to `LOG_ERR("bus %s is not ready", bq34110_cfg->i2c_dev->name)` or so.